src/validation.cpp

```diff
@@ -3075,6 +3075,232 @@ static bool CheckBlockHeader(const CBlockHeader& block, CValidationState& state,
+     * collisions)
+     *
+     * For each of the 8 hashes h_1...h_8, P(bit set in table for h_i) =
+     * (N*8)/1<<21
+     *
+     * Each of these probabilities is independent
+     *
+     * Therefore the total probability of a false collision on all bits is:
+     * ((N*8)/2**21)**8
+     *
+     * The cost of a false collision is to do N comparisons.
+     *
+     * Therefore, the expression for the expected number of comparisons is:
+     *
+     * Sum[i*( i*8 / 2**21)**8, {i, 0, M}]
+     *
+     * Based on an input being at least 41 bytes, and a block being 1M bytes
+     * max, there are a maximum of 24390 inputs, so M = 24390
+     *
+     * The total expected number of direct comparisons for M=24930 is therefore
+     * 0.33 with this algorithm.
+     *
+     * The worst case for this algorithm from a denial of service perspective
+     * with an invalid block would be to do a transaction where the last two
+     * elements are a collision. In this case, the scan would require to scan all
+     * N elements to find the conflict.
+     *
+     * ------------------------------------------------------------------------
+     *  Extensions
+     * ------------------------------------------------------------------------
+     *
+     * - Single Coinbase / Null Input Check
+     *     Note that the first element checked is the coinbase transaction,
+     *     whose input is null. Therefore, any subsequent null input would be a
+     *     collision with that null, enabling us to not null check every
+     *     subsequent entry. This has 0 overhead.
+     * - Long Chain Check
+     *     We also scan for the presence of the outputs of a transaction in the
+     *     table as we go (without insertion), which detects an out-of-order
+     *     spend in a long-chain within a block.
+     *
+     *     The worst case behavior for a block under this additional scan is for
+     *     all inputs to be spent and then all outputs to be created. Any other
+     *     pattern of inputs and outputs would be strictly less work. A minimal
+     *     output is 9 bytes -- there are at most 1e6/9 outputs. Thus, we can
+     *     model it as:
+     *
+     *     Max[((1-x)*1e6/9)(x*1e6/41)*(8*x*(1e6/41) / 2**21)**8 +
+     *         Sum[ i*( i*8 / 2**21)**8, {i, 0, x*1e6/41}], {x, 0, 1}]
+     *
+     *     That is:
+     *       The most amount of work for a given fraction x of block space devoted to
+     *       inputs, which is the expected amount of work for checking a table with
+     *       x*1e6/41 entries (1-x)*1e6/9 additional times plus the expected  work
+     *       for deduplicating just the inputs.
+     *
+     *     This expression is at most ~0.7 expected comparisons worst case, which is
+     *     still perfectly acceptable.
+     */
+
+
+    const auto pcg = [](uint64_t start, uint64_t inc) {
+        uint64_t nxt = (inc | 1) + (start * 6364136223846793005ULL);
```

Comment:
This is an unsigned integer overflow. Make it explicit that this unsigned integer overflow is intentional?

Reply:
How do I make the wraparound explicit?

Reply:
You could use the method I used in PR #14224: 

https://github.com/practicalswift/bitcoin/blob/0373038cedc07e3d775a1055c0da45f4e8fda2b7/src/attributes.h#L9-L15

Reply:
Why would that be preferable to the sanitizer suppressions in `./test/`?

Reply:
Yeah I'm not sure about it. Unsigned integer overflow isn't a bug, it's a feature... I would also like a suppression that operates at the statement level (e.g., like rust's wrapping_mul) rather than marking the entire function with an attribute. What if there is an unsigned wraparound issue elsewhere in the code?

Reply:
@MarcoFalke In contrast to the sanitizer suppressions it documents if the wraparound is intentional or not.